File: 76_Minimum_Window_Substring/minWindow.py
class Solution(object):
    def minWindow(self, s, t):
        dict_map = {c:-1 for c in t}
        res = s
        for idx, c in enumerate(s):
            #print dict_map, res
            if c in dict_map:
                dict_map[c] = idx
            if -1 not in dict_map.values():
                res = min(s[min(dict_map.values()):idx+1], res, key=len)
                dict_map[s[min(dict_map.values())]] = -1
            #print dict_map, res
        #print res
        return res

sol = Solution()
s = "ADOBECODEBANC"
t = "ABC"
print s
sol.minWindow(s, t)

File: 76_Minimum_Window_Substring/minWindow.cpp
#include <iostream>
#include <string>
#include <map>

using namespace std;

class Solution {
	public:
    	string minWindow(string s, string t) {
            map<char, int> need;
            size_t i = 0, I = 0, J = 0;
            for(size_t k = 0; k < t.size(); k++)
                need[t[k]]++;

            size_t missing = t.size();
            for(size_t j = 0; j < s.size(); j++) {
                missing -= size_t(need[s[j]] > 0);
                need[s[j]]--;
                if(!missing) {
                    while(i < (j+1) && need[s[i]] < 0) {
                        need[s[i]]++;
                        i++;
                    }
                    if(!J || (j+1) - i <= J - I) {
                        I = i;
                        J = j + 1;
                    }
                }
            }
            return s.substr(I, J - I);
    	}
};

int main() {
    Solution sol;
    string s = "a";
    string t = "a";
    cout << sol.minWindow(s, t) << endl;
}
